### ISSUE TITLE
ansible-playbook v2.4 slow performance  - compared to v2.3.2 

### ISSUE THREAD

**Issue by PhilEv** [ansible/ansible#30519]
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

ansible-playbook 

##### ANSIBLE VERSION

```
ansible-playbook --version
ansible-playbook 2.4.0.0 (detached HEAD 3c805f85bc) last updated 2017/09/18 15:49:22 (GMT +200)
  config file = /home/eveque/.ansible.cfg
  configured module search path = [u'/home/eveque/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/eveque/workspace/ansible/lib/ansible
  executable location = /home/eveque/workspace/ansible/bin/ansible-playbook
  python version = 2.7.13 (default, Sep  5 2017, 08:53:59) [GCC 7.1.1 20170622 (Red Hat 7.1.1-3)]

```

##### CONFIGURATION

```ini
[defaults]

# some basic default values...

inventory       = /home/eveque/workspace/inventories/test_infra
callback_plugins   = /home/eveque/workspace/tools4ct/callback_plugins
timeout = 30

log_path = /home/eveque/ansible.log
force_color = 1

retry_files_enabled = False

[ssh_connection]

ssh_args = -o ControlMaster=auto -o ControlPersist=60s
control_path = %(directory)s/ansible-ssh-%%h-%%p-%%r
pipelining = True
```
##### OS / ENVIRONMENT

Ansible controle node OS : Fedora26
Managed system OS:  Red Hat Enterprise Linux Server release 7.4 (Maipo)

##### SUMMARY

Running the same playbook with the same set of options, inventory file, callback  and same configuration  results in significant performance between ansible 2.3.2 and 2.4.0.rc4 (or rc5)

The time for running  this playbook is nearly twice as much as with ansible 2.3.2

##### STEPS TO REPRODUCE

no minimal test case yet

with ansible 2.4.0 after sourcing the `hacking/env-setup`

```
ansible-playbook --version
ansible-playbook 2.4.0.0 (detached HEAD 3c805f85bc) last updated 2017/09/18 15:49:22 (GMT +200)
  config file = /home/eveque/.ansible.cfg
  configured module search path = [u'/home/eveque/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/eveque/workspace/ansible/lib/ansible
  executable location = /home/eveque/workspace/ansible/bin/ansible-playbook
  python version = 2.7.13 (default, Sep  5 2017, 08:53:59) [GCC 7.1.1 20170622 (Red Hat 7.1.1-3)
```
then
```
time ansible-playbook ocoe.yml -Kv -i ~/workspace/inventories/one_host  -u tester -e targets="helios5.gre.hpecorp.net" 

Using /home/eveque/.ansible.cfg as config file
SUDO password: 
[DEPRECATION WARNING]: 'include' for playbook includes. You should use 'import_playbook' instead. This feature will be removed in version 2.8. Deprecation warnings can be disabled by setting 
deprecation_warnings=False in ansible.cfg.
[DEPRECATION WARNING]: The use of 'include' for tasks has been deprecated. Use 'import_tasks' for static inclusions or 'include_tasks' for dynamic inclusions. This feature will be removed in a future release. 
Deprecation warnings can be disabled by setting deprecation_warnings=False in ansible.cfg.

PLAY [Basic OE configuration tuning] *******************************************************************************************************************************************************************************
...

PLAY RECAP *********************************************************************************************************************************************************************************************************
helios5.gre.hpecorp.net    : ok=30   changed=3    unreachable=0    failed=0   

real	0m25.885s
user	0m5.275s
sys	0m3.566s

```
running with the version 2.3.2

```
rpm -qa ansible
ansible-2.3.2.0-1.fc26.noarch
```
Same configuration file and callback and inventory as above
```
ansible-playbook 2.3.2.0
  config file = /home/eveque/.ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.13 (default, Sep  5 2017, 08:53:59) [GCC 7.1.1 20170622 (Red Hat 7.1.1-3)]

```
then
```
time ansible-playbook ocoe.yml -Kv -i ~/workspace/inventories/one_host  -u tester -e targets="helios5.gre.hpecorp.net" 
Using /home/eveque/.ansible.cfg as config file
SUDO password: 

PLAY [Basic OE configuration tuning] *******************************************************************************************************************************************************************************
...
PLAY RECAP *********************************************************************************************************************************************************************************************************
helios5.gre.hpecorp.net    : ok=30   changed=3    unreachable=0    failed=0   

real	0m14.410s
user	0m3.073s
sys	0m1.543s
```

##### EXPECTED RESULTS

similar timing for both run

##### ACTUAL RESULTS

v 2.4.0rc4 took almost twice the time to run when compared to 2.3.2 

```

```

**Comment by jctanner** [comment]
@PhilEv It's difficult to make an assessment here without an example playbook. There are a few areas where the performance could have degraded:

1. inventory
2. connection
3. privledge escalation
4. the module(s) befing called

Please provide a reproducer playbook or use ANSIBLE_DEBUG=1 to get timestamps for each phase and narrow down where the slowness occurs by comparing the output between both versions.

needs_info

**Comment by PhilEv** [comment]
@jctanner I agree that is required and will provide the info.
Note the inventory was one file with one host.  More to come.

**Comment by PhilEv** [comment]
here are the outputs of 2 runs (one with 2.3.2 and one with 2.4.0)
This is mosty using the lineinfile(mostly), template, replace, cmd, shell,file, stat

[oetuneoutput_ansible232.txt](https://github.com/ansible/ansible/files/1316548/oetuneoutput_ansible232.txt)
[oetuneoutput_ansible240.txt](https://github.com/ansible/ansible/files/1316549/oetuneoutput_ansible240.txt)

will check if I can provide the playbook and roles (nothing secret, but I have to)

**Comment by PhilEv** [comment]
@jctanner 

here are some chart showing the distribution
[oetune_output.xlsx](https://github.com/ansible/ansible/files/1316600/oetune_output.xlsx)

metrics got via 
```
ag 'running TaskExecutor' oetuneoutput_ansible232.txt
```
and similar with the other file

**Comment by towolf** [comment]
Don't want to divert attention in this issue, but inventory parsing time for our large inventory increased 4 fold:

- `--list-hosts` with 2.2: 5 sec
- `--list-hosts` with 2.4: 20 sec

**Comment by bcoca** [comment]
so this PR #30587 speeds up inventory to 2.3/even faster speeds, let me know if this solves the bulk of the issues

**Comment by PhilEv** [comment]
@bcoca , I guess this will help/fix the @towolf issue. In my case the inventory was made of 1 host in 2 groups.
I will give it a try anyway.

**Comment by towolf** [comment]
One issue down, yes. I have not had the chance to test properly, yet. But one playbook run yesterday running stable-2.4 from the git checkout seemed really slow, with ok/skip tasks ticking by very slowly ...

**Comment by towolf** [comment]
https://www.reddit.com/r/ansible/comments/71nusw/24_much_slower_than_23/

**Comment by PhilEv** [comment]
@bcoca and @jctanner After reading issue #31125 I tested again with setting the `ANSIBLE_PIPELINING` env var to true. The timing is then much closer to the one of Ansible 2.3 and then Ok for me.

So a confirmation for PIPELINING not being active by default (as reported in issue #31125) is the main source of perf degradation in my case too.


here are the last round of results:
 
```
ansible-playbook 2.4.1.0 (detached HEAD 085297259f) last updated 2017/09/28 10:44:51 (GMT +200)
  config file = /home/eveque/.ansible.cfg
  configured module search path = [u'/home/eveque/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/eveque/workspace/ansible/lib/ansible
  executable location = /home/eveque/workspace/ansible/bin/ansible-playbook
  python version = 2.7.13 (default, Sep  5 2017, 08:53:59) [GCC 7.1.1 20170622 (Red Hat 7.1.1-3)]

export ANSIBLE_PIPELINING=True
time ansible-playbook ocoe.yml -Kv -i ~/workspace/inventories/one_host  -u tester -e targets="helios5.gre.hpecorp.net"
Using /home/eveque/.ansible.cfg as config file
SUDO password: 

PLAY [Basic OE configuration tuning] ********************************************************************************************************************************************************************************************************

...


PLAY RECAP **********************************************************************************************************************************************************************************************************************************
helios5.gre.hpecorp.net    : ok=30   changed=3    unreachable=0    failed=0   


real	0m16.068s
user	0m5.012s
sys	0m1.557s
```

When the  ANSIBLE_PIPELINING env var is unset, the timing are back to the original observation

```
unset ANSIBLE_PIPELINING
ansible-playbook --version
ansible-playbook 2.4.1.0 (detached HEAD 085297259f) last updated 2017/09/28 10:44:51 (GMT +200)
  config file = /home/eveque/.ansible.cfg
  configured module search path = [u'/home/eveque/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/eveque/workspace/ansible/lib/ansible
  executable location = /home/eveque/workspace/ansible/bin/ansible-playbook
  python version = 2.7.13 (default, Sep  5 2017, 08:53:59) [GCC 7.1.1 20170622 (Red Hat 7.1.1-3)]

time ansible-playbook ocoe.yml -Kv -i ~/workspace/inventories/one_host  -u tester -e targets="helios5.gre.hpecorp.net"
Using /home/eveque/.ansible.cfg as config file
SUDO password: 

PLAY [Basic OE configuration tuning] ********************************************************************************************************************************************************************************************************

...

PLAY RECAP **********************************************************************************************************************************************************************************************************************************
helios5.gre.hpecorp.net    : ok=30   changed=3    unreachable=0    failed=0   


real	0m26.629s
user	0m6.165s
sys	0m3.226s

**Comment by bcoca** [comment]
then #31244 should restore performance w/o requiring a configuration change.

**Comment by PhilEv** [comment]
@bcoca I confirm it is solved on the stable-2.4 branch after testing again 

getting now:

```
...
PLAY RECAP **************************************************************************************************************************************************************************************************************
helios5.gre.hpecorp.net    : ok=30   changed=3    unreachable=0    failed=0   


real	0m16.895s
user	0m4.944s
sys	0m1.587s
```

with the version below and with the same configursation used with 2.3.x

```
ansible-playbook 2.4.1.0 (stable-2.4 43d24e284c) last updated 2017/10/04 11:09:49 (GMT +200)
  config file = /home/eveque/.ansible.cfg
  configured module search path = [u'/home/eveque/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/eveque/workspace/ansible/lib/ansible
  executable location = /home/eveque/workspace/ansible/bin/ansible-playbook
  python version = 2.7.13 (default, Sep  5 2017, 08:53:59) [GCC 7.1.1 20170622 (Red Hat 7.1.1-3)]
```

I agree to close it
thanks

**Comment by bcoca** [comment]
closing as per above.
